tests/unit/stall_detector_test.cc

```diff
@@ -106,3 +138,42 @@ SEASTAR_THREAD_TEST_CASE(no_poll_no_stall) {
     f.get();
     BOOST_REQUIRE_EQUAL(reports, 0);
 }
+
+// Triggers stalls by spinning with a specify "body" function
+// which takes most of the spin time. The tests below that use
+// this do not assert anything, but are intended for diagnosing
+// stall problems by expecting the output. We expect the userspace
+// spin test to show no kernel callstack, and the kernel test to
+// show kernel backtraces in the mmap or munmap path, but this is 
+// not exact since neither test spends 100% of its time in the 
+// selected mode.
+template <typename B>
+static void test_spin_with_body(const char* what, B body) {
+    testlog.info("Starting spin test: {}", what);
+    temporary_stall_detector_settings tsds(10ms);
+    constexpr unsigned nr = 5;
+    for (unsigned i = 0; i < nr; ++i) {
+        spin_some_cooperatively(100ms, body);
+        spin(20ms, body);
+    }
+    testlog.info("Ending spin test: {}", what);
+}
+
+SEASTAR_THREAD_TEST_CASE(spin_in_userspace) {
+    // a body which spends almost all of its time in userspace
+    test_spin_with_body("userspace", [] { spin_user_hires(1ms); });
+}
+
+static void mmap_populate(size_t len) {
+    void *p = mmap(nullptr, len, PROT_READ | PROT_WRITE, MAP_PRIVATE | MAP_ANONYMOUS | MAP_POPULATE, 0, 0);
+    BOOST_REQUIRE(p != MAP_FAILED);
+    BOOST_REQUIRE(munmap(p, len) == 0);
+}
+
+SEASTAR_THREAD_TEST_CASE(spin_in_kernel) {
+    // a body which spends almost all of its time in the kernel
+    // doing 128K mmaps
+    test_spin_with_body("kernel", [] { mmap_populate(128 * 1024); });
```

Comment:
Why not assert some stuff? Nobody will run the test and check the results are okay.

Reply:
The idea is for the actual output to stderr you need something to hack against while you are fixing this, so while I don't expect anyone to run this randomly, I would like it there for the next time I come back to work on this and it may help someone else who has to do the same thing. Asserting some things I would like to would make the test flappy.

That said, I've modified it now to run both in output mode and assert mode and in the latter it checks that the right number of stacks are reported.